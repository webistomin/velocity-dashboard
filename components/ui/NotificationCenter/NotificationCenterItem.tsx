import { VNode } from 'vue';
import { BaseIcon } from 'components/base/BaseIcon/BaseIcon';
import { VueComponent } from 'types/vue-components';
import { BaseNotify } from 'components/base/BaseNotify/BaseNotify';
import { Component, Prop } from '~/node_modules/nuxt-property-decorator';

export enum NotificationCenterItemTypes {
  MESSAGE = 'message',
  APPLICATION = 'application',
  COMPLAINT = 'complaint',
}

export interface INotificationCenterItemProps {
  id: string | number;
  type: NotificationCenterItemTypes;
  text: string;
  time: DOMTimeStamp;
  isNew: boolean;
}

@Component({
  name: 'NotificationCenterItem',
})
export default class NotificationCenterItem extends VueComponent<INotificationCenterItemProps> {
  @Prop({ default: NotificationCenterItemTypes.APPLICATION })
  private readonly type!: INotificationCenterItemProps['type'];

  @Prop({ default: '' })
  private readonly text!: INotificationCenterItemProps['text'];

  @Prop({ default: '' })
  private readonly time!: INotificationCenterItemProps['time'];

  @Prop({ default: '' })
  private readonly id!: INotificationCenterItemProps['id'];

  @Prop({ default: false })
  private readonly isNew!: INotificationCenterItemProps['isNew'];

  get getIconName() {
    switch (this.type) {
      case NotificationCenterItemTypes.MESSAGE:
        return 'icon-bubble';
      case NotificationCenterItemTypes.APPLICATION:
        return 'icon-mail';
      case NotificationCenterItemTypes.COMPLAINT:
        return 'icon-lightning';
      default:
        return 'icon-mail';
    }
  }

  render(): VNode {
    return (
      <div class='notification-center__item'>
        <nuxt-link to='/test' class='notification-center__item-link link'>
          <div class='notification-center__col notification-center__col_left'>
            {this.isNew ? <BaseNotify/> : null}
            <BaseIcon
              class={`notification-center__item-icon ${this.isNew ? 'notification-center__item-icon_new' : ''}`}
              name={this.getIconName}
              size='xs'
            />
          </div>
          <div class='notification-center__col notification-center__col_right'>
            <strong class='notification-center__caption caption'>{this.type.toUpperCase()}</strong>
            <p class='notification-center__text paragraph'>{this.text}</p>
            <time class='notification-center__time'>{this.time}</time>
          </div>
        </nuxt-link>
      </div>
    );
  }
}
