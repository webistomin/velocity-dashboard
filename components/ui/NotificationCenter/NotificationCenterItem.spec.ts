import { shallowMount } from '@vue/test-utils';
import NotificationCenterItem, { NotificationCenterItemTypes } from './NotificationCenterItem';

describe('NotificationCenterItem', () => {
  const factory = (propsData = {}, options = {}) => {
    return shallowMount(NotificationCenterItem, {
      propsData: {
        type: NotificationCenterItemTypes.MESSAGE,
        text: '',
        time: 1591086090152,
        id: '1',
        isNew: true,
        ...propsData,
      },
      stubs: ['nuxt-link'],
      ...options,
    });
  };

  it('Matches snapshot', () => {
    const wrapper = factory().html();
    expect(wrapper).toMatchSnapshot();
  });

  it('Return "icon-bubble" for message notification', () => {
    const wrapper = factory({
      type: NotificationCenterItemTypes.MESSAGE,
    });
    const result = wrapper.vm.getIconName;
    expect(result).toEqual('icon-bubble');
  });

  it('Return "icon-mail" for application notification', () => {
    const wrapper = factory({
      type: NotificationCenterItemTypes.APPLICATION,
    });
    const result = wrapper.vm.getIconName;
    expect(result).toEqual('icon-mail');
  });

  it('Return "icon-lightning" for complaint notification', () => {
    const wrapper = factory({
      type: NotificationCenterItemTypes.COMPLAINT,
    });
    const result = wrapper.vm.getIconName;
    expect(result).toEqual('icon-lightning');
  });

  it('Return "icon-mail" by default', () => {
    const wrapper = factory({
      type: '',
    });
    const result = wrapper.vm.getIconName;
    expect(result).toEqual('icon-mail');
  });
});
