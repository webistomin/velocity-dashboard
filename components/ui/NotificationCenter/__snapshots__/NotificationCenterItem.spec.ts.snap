// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationCenterItem Matches snapshot 1`] = `
"<div class=\\"notification-center__item\\">
  <nuxt-link-stub to=\\"#\\" class=\\"notification-center__item-link link\\">
    <div class=\\"notification-center__col notification-center__col_left\\">
      <vuecomponent-stub></vuecomponent-stub>
      <vuecomponent-stub size=\\"xs\\" name=\\"icon-bubble\\" color=\\"default\\" class=\\"notification-center__item-icon notification-center__item-icon_new\\"></vuecomponent-stub>
    </div>
    <div class=\\"notification-center__col notification-center__col_right\\"><strong class=\\"notification-center__caption caption\\">MESSAGE</strong>
      <p class=\\"notification-center__text paragraph\\"></p><time class=\\"notification-center__time\\">1591086090152</time></div>
  </nuxt-link-stub>
</div>"
`;
